File: src/main/java/core/procedures/strings/Join.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.utils.Utils;

import java.util.Iterator;

public final class Join extends AFn {

  public Join() {
    super(new FnArgsBuilder().min(1).max(2).mandatory(new Class[]{CharSequence.class}).build());
  }

  @Override
  public String getName() {
    return "join";
  }

  @Override
  public String apply(Object... args) {
    if (args.length == 1) {
      return args[0].toString();
    }
    String separator = args[0].toString();
    Iterator iterator = Utils.toIterator(args[1]);
    if (!iterator.hasNext()) {
      return "";
    }
    StringBuilder sb = new StringBuilder();
    while (iterator.hasNext()) {
      sb.append(iterator.next());
      if (iterator.hasNext()) {
        sb.append(separator);
      }
    }
    return sb.toString();
  }
}

File: src/main/java/core/procedures/strings/Join.kt
package core.procedures.strings

import core.Writer
import core.procedures.AFn
import core.procedures.Arity.Range
import core.utils.Utils

class Join : AFn<Any?, String>(name = "join", isPure = true, arity = Range(1, 2),
                               mandatoryArgsTypes = arrayOf(CharSequence::class.java)) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        1    -> args[0].toString()
        else -> Utils.toSequence(args[1]).joinToString(separator = args[0].toString()) { str(it) }
    }

    private fun str(obj: Any?) = when (obj) {
        null -> ""
        is Char -> obj.toString()
        is CharSequence -> obj
        else -> Writer.write(obj)
    }
}
